src/app-layer-register.c

```diff
@@ -0,0 +1,191 @@
+/* Copyright (C) 2017 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+/**
+ * \file
+ *
+ * \author Pierre Chifflier <chifflier@wzdftpd.net>
+ *
+ * Parser registration functions.
+ */
+
+#include "suricata-common.h"
+#include "stream.h"
+#include "conf.h"
+
+#include "app-layer-detect-proto.h"
+#include "app-layer-parser.h"
+
+#include "app-layer-register.h"
+
+static const char * IpProtoToString(int ip_proto);
+
+AppProto AppLayerRegisterProtocolDetection(const struct AppLayerParser *p, int enable_default)
+{
+    AppProto alproto;
+    const char *ip_proto_str = NULL;
+
+    if (p == NULL)
```

Comment:
this should probably print an error or simply abort()/BUG_ON as this is a coding error, not a user config error

Reply:
I copied the behavior from `app-layer-parser.c`: these errors are now fatal (they can only result from a programming mistake and cannot be recovered), they print an error and exit using
`SCLogError(SC_ERR_FATAL, ...) + exit(EXIT_FAILURE)`.

Since this function can now only return a valid `AppProto`, I update the inline documentation and simplified the rust calls.

Reply:
You can use FatalError, it takes the same args as SCLogError but exits

Reply:
I see, the macro does both things. Done!